File: Z1.1/client1_1.py
import socket
import sys
import argparse


def end(s):
    print("end of work")
    s.close()


def send_data(s, msg):
    try:
        s.sendto(str.encode(msg), (HOST, PORT))
        print(f"sent message len = {len(msg)}")
    except WindowsError as e:
        return -2, len(msg)
    response = ""
    while not response:
        try:
            response, server_addr = s.recvfrom(1024)
            print(f"Server response: {response!r}")
            if int(response) != len(msg):
                print("verification error")
                end(s)
            return 0, int(response)
        except socket.timeout:
            print("socket timeout")
            end(s)


def setup():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((HOST, PORT))
    print("created socket")
    x = 2
    x_o = -1
    a = 2
    r = 0
    s.settimeout(10)
    while abs(x -x_o)>1:
        msg = "A"*x
        if x >=3:
            msg = "z32" + "A"*(x-3)
        res, r = send_data(s, msg)
        if(x==r):
            if res == -2:
                x = int(x - (x- x_o)/2)
                continue
            else:
                x_o = x
                x = int(x*a)
    end(s)


def main(arguments):
    parser = argparse.ArgumentParser()
    parser.add_argument("--IP", type=str)
    parser.add_argument("--PORT", type=int)
    args = parser.parse_args(arguments[1:])
    global HOST
    global PORT
    if not args.IP:
        HOST = "127.0.0.1"
    else:
        HOST = args.IP
    if not args.PORT:
        PORT = 3201
    else:
        PORT = args.PORT
    setup()


if __name__ == "__main__":
    main(sys.argv)

File: Z1.1/client1_1.cpp
// Client side implementation of UDP client-server model
#include <arpa/inet.h>
#include <bits/stdc++.h>
#include <netinet/in.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

#define PORT 8080
#define MAXLINE 66000

char *msg_generator(int lenght) {
  char *message = new char[66000];
  char sender[] = "z32";
  strncpy(message, sender, sizeof(sender));
  for (int i = 3; i < lenght; i++) {
    message[i] = char(i % 26 + 65);
  }
  message[lenght] = '\0';
  return message;
}

// Driver code
int main() {
  int sockfd;
  char buffer[MAXLINE];
  // const char *hello = "Hello from client";
  struct sockaddr_in servaddr;

  // Creating socket file descriptor
  if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    perror("socket creation failed");
    exit(EXIT_FAILURE);
  }

  memset(&servaddr, 0, sizeof(servaddr));

  // Filling server information
  servaddr.sin_family = AF_INET;
  servaddr.sin_port = htons(PORT);
  servaddr.sin_addr.s_addr = INADDR_ANY;

  int n;
  socklen_t len;
  int length = 1;

  while (1) {
    char *msg = msg_generator(length);

    sendto(sockfd, msg, strlen(msg), MSG_CONFIRM,
           (const struct sockaddr *)&servaddr, sizeof(servaddr));
    std::cout << "Message sent with length: " << length << std::endl;

    length += 100;
    n = recvfrom(sockfd, (char *)buffer, MAXLINE, MSG_WAITALL,
                 (struct sockaddr *)&servaddr, &len);
    buffer[n] = '\0';
    std::cout << "Server received message:" << buffer << std::endl;
  }

  close(sockfd);
  return 0;
}
